Libs/RaptureAddins/ZooKeeper/src/main/java/rapture/lock/zookeeper/ZooKeeperLockHandler.java

```diff
@@ -0,0 +1,206 @@
+
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.CuratorFrameworkFactory;
+import org.apache.curator.framework.recipes.locks.InterProcessSemaphoreMutex;
+import org.apache.curator.framework.state.ConnectionState;
+import org.apache.curator.framework.state.ConnectionStateListener;
+import org.apache.curator.retry.RetryUntilElapsed;
+import org.apache.log4j.Logger;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.data.Stat;
+
+import rapture.common.LockHandle;
+import rapture.common.exception.ExceptionToString;
+import rapture.config.MultiValueConfigLoader;
+import rapture.lock.ILockingHandler;
+
+/**
+ * An implementation of a lock strategy on ZK.
+ * { name = lockName, locked=true, freeTime=(some long time) }
+ * 
+ * @author James Howe
+ * 
+ */
+public class ZooKeeperLockHandler implements ILockingHandler {
+
+    @Override
+    public String toString() {
+        return "ZooKeeperLockHandler [locks=" + locks + ", client=" + client + ", instanceName=" + instanceName + ", config=" + config + "]";
+    }
+
+    private static Logger logger = Logger.getLogger(ZooKeeperLockHandler.class);
+    private static final int TIME_TO_WAIT_DEFAULT = 3;
+    private static final String LOCK_BASE = "/rapture_locking";
+
+    private Map<String, InterProcessSemaphoreMutex> locks;
+
+    // TODO put this mutex in a map with the key being lockName...
+    // InterProcessSemaphoreMutex lock;
+
+    CuratorFramework client;
+    String instanceName = "dunno";
+    Map<String, String> config = null;
+
+    public ZooKeeperLockHandler(String connectionString, int timeToWait) {
+        locks = new HashMap<>();
+        if (timeToWait == 0) {
+            logger.info("time to wait not set. setting to default: " + TIME_TO_WAIT_DEFAULT);
+            timeToWait = TIME_TO_WAIT_DEFAULT;
+        }
+
+        logger.debug("connectionString is " + connectionString);
+        logger.debug("timeToWait is " + timeToWait);
+
+        client = CuratorFrameworkFactory.newClient(connectionString, new RetryUntilElapsed(3, 250));
+
+        client.getConnectionStateListenable().addListener(new ConnectionStateListener() {
+            @Override
+            public void stateChanged(CuratorFramework client, ConnectionState newState) {
+                if (newState == ConnectionState.LOST) {
+                    logger.error("lost connection to zookeeper server");
+                    // TODO connection CAN'T be trusted anymore. throw a runtime telling the client app to reconnect.
+                } else if (newState == ConnectionState.SUSPENDED) {
+                    logger.error(" connection suspended to zookeeper server");
+                    // TODO we can wait for state to move back to CONNECTED. if not time out with exception
+                }
+            }
+        });
+        client.start();
+    }
+
+    public ZooKeeperLockHandler() {
+        this("localhost:2181", 0);
```

Comment:
That is the default, but you can specify any host and port by using the other constructor.